README.md
# Elasticsearch Partition
[![image](https://img.shields.io/pypi/v/elasticsearch-partition.svg)](https://pypi.python.org/pypi/elasticsearch-partition)
[![Build Status](https://travis-ci.com/dmvass/elasticsearch-partition.svg?branch=master)](https://travis-ci.com/kandziu/elasticsearch-partition)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/dmvass/elasticsearch-partition/blob/master/LICENSE)

A Python library is written on Cython for creating Elasticsearch indexes by
date range.

For time oriented data, such as logs, a common strategy is to partition data
into indexes that hold data for a certain time range. For example, the index
`logstash-2018.01.01` holds data for events that happened on `2018-01-01`, i.e.
a time range of a day. You can of course choose bigger or smaller time ranges
as well(`year`, `month` or `day` frequencies), depending on your needs. Using
index templates, you can easily manage settings and mappings for any index
created with a name starting with e.g. `logstash-*`.

## Installation
Install the elasticsearch partition package with pip:
```
pip install elasticsearch-partition
```

## How to Use
### Basic usage
How to import and use partition module with `since` and `until` dates:
```python
import datetime
from elasticsearch_partition import partition

partition('logs-*', datetime.date(2016, 11, 29), datetime.date(2018, 2, 4))
# ['logs-2016-11-29', 'logs-2016-11-30', 'logs-2016-12-*', 'logs-2017-*',
# 'logs-2018-01-*', 'logs-2018-02-01', 'logs-2018-02-02', 'logs-2018-02-03',
# 'logs-2018-02-04']
```

When you are using `partition` only with `since` date, `until` will be replaced
on a current date.
```python
partition('logs-*', since=datetime.date(2018, 7, 10))
# ['logs-2018-07-10', 'logs-2018-07-11', 'logs-2018-07-12', 'logs-2018-07-13',
# 'logs-2018-07-14', 'logs-2018-07-15', 'logs-2018-07-16', 'logs-2018-07-17']
```

Or when you are using `partition` only with `until` all dates from `until` to
current date will be excluded.
```python
partition('logs-*', until=datetime.date(2018, 7, 10))
# ['-logs-2018-07-10', '-logs-2018-07-11', '-logs-2018-07-12',
# '-logs-2018-07-13', '-logs-2018-07-14', '-logs-2018-07-15',
# '-logs-2018-07-16', '-logs-2018-07-17', 'logs-*']
```

> Note: If `until` more then current date you will get an error.

### How to customize partitioning
If you want to change some `partition` bahavior you can do it ease with
`RangePartition` and `formatters` module, also you can use your custom date
`now` functions.
```python
from elasticsearch_partition import RangePartition
from elasticsearch_partition.partitioning import MONTH
from elasticsearch_partition.formatters import LittleEndianDateFormatter

# frequency - Index partitioning frequency
# formatter - Formatter instance
# escape - Special character which will be replaced on a date
# now_func - Get now date function
my_partition = RangePartition(
    frequency=MONTH,
    formatter=LittleEndianDateFormatter(sep='.'),
    escape='@',
    now_func=custom_date_now,
)

my_partition('logs-@', datetime.date(2016, 11, 29), datetime.date(2018, 2, 4))
# ['logs-11.2016', 'logs-12.2016', 'logs-*.2017', 'logs-01.2018', 'logs-02.2018']
```

### How to create custom date formatter
All date formatters must be inherited from abstract `DateFormatter` class and
implement `fmt_year`, `fmt_month` and `fmt_day` methods. Some method accept
additional keyword parameter `wildcard` which used for creating formatted date
with specified wildcard character. For example `2018-04` will be replced on
`2018-04-*`, `2018` on `2018-*` etc.
```python
class MyDateFormatter(DateFormatter):
    def fmt_year(self, year, wildcard):
        # Should be implemented

    def fmt_month(self, year, month, wildcard):
        # Should be implemented
    
    def fmt_day(self, year, month, day):
        # Should be implemented

partition = RangePartition(formatter=MyDateFormatter())
```

### How to use with [elasticsearch-py](https://github.com/elastic/elasticsearch-py)
This is useful for all Elasticsearch APIs that refer to an index parameter
support execution across multiple indices.
```python
from elasticsearch import Elasticsearch

es = Elasticsearch()
indexes = partition(
    'logs-*',
    datetime.date(2016, 11, 29),
    datetime.date(2018, 2, 4)
)
res = es.search(index=indexes, body={"query": {"match_all": {}}})
```

### How to use with [elasticsearch-dsl-py](https://github.com/elastic/elasticsearch-dsl-py)
This is useful for all Elasticsearch APIs that refer to an index parameter
support execution across multiple indices and similar for simple Search and
Persistance DSL.
```python
from elasticsearch import Elasticsearch
from elasticsearch_dsl import Search

client = Elasticsearch()

indexes = partition(
    'logs-*',
    datetime.date(2016, 11, 29),
    datetime.date(2018, 2, 4)
)
search = Search(using=client, index=indexes) \
    .filter("term", category="search") \
    .query("match", title="python") \
    .exclude("match", description="beta")

response = search.execute()
```

## Changes
A full changelog is maintained in the [CAHNGELOG](https://github.com/dmvass/elasticsearch-partition/blob/master/CHANGELOG.md) file.

## Contributing 
**elasticsearch-partition** is an open source project and contributions are
welcome! Check out the [Issues](https://github.com/dmvass/elasticsearch-partition/issues)
page to see if your idea for a contribution has already been mentioned, and feel
free to raise an issue or submit a pull request.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Makefile

## [2.0.0] - 2019-03-22
### Added
- Cython implementation

## [1.0.3] - 2019-02-23
### Fixed
- Package description content type

## [1.0.2] - 2019-02-23
### Added
- Code style and linting tox environment
- PyPy and Cython support
- Cython compilation
### Fixed
- little-endian formatter fmt_month with wildcard

## [1.0.1] - 2018-07-19
### Added
- Validation index name separation

## [1.0.0] - 2018-07-17
- Initial stable release.

tests/test_formatters.py
import unittest

from elasticsearch_partition import formatters


class TestBigEndianDateFormatter(unittest.TestCase):

    def setUp(self):
        self.module = formatters.BigEndianDateFormatter
        self.formatter = self.module()

    def test_fmt_year(self):
        # Test without wildcard
        actual = self.formatter.fmt_year(2018, wildcard=False)
        self.assertEqual(actual, "2018")
        # Test with wildcard
        actual = self.formatter.fmt_year(2018, wildcard=True)
        self.assertEqual(actual, "2018-*")

    def test_fmt_month(self):
        # Test without wildcard
        actual = self.formatter.fmt_month(2018, 4, wildcard=False)
        self.assertEqual(actual, "2018-04")
        # Test with wildcard
        actual = self.formatter.fmt_month(2018, 4, wildcard=True)
        self.assertEqual(actual, "2018-04-*")

    def test_fmt_day(self):
        actual = self.formatter.fmt_day(2018, 4, 20)
        self.assertEqual(actual, "2018-04-20")

    def test_sep(self):
        formatter = self.module(sep=".")
        actual_year = formatter.fmt_year(2018, wildcard=False)
        actual_month = formatter.fmt_month(2018, 4, wildcard=False)
        actual_day = formatter.fmt_day(2018, 4, 20)
        self.assertEqual(actual_year, "2018")
        self.assertEqual(actual_month, "2018.04")
        self.assertEqual(actual_day, "2018.04.20")

        for sep in formatters.INVALID_SEP_CHARS:
            with self.assertRaises(ValueError):
                self.module(sep)


class TestLittleEndianDateFormatter(unittest.TestCase):

    def setUp(self):
        self.module = formatters.LittleEndianDateFormatter
        self.formatter = self.module()

    def test_fmt_year(self):
        # Test without wildcard
        actual = self.formatter.fmt_year(2018, wildcard=False)
        self.assertEqual(actual, "2018")
        # Test with wildcard
        actual = self.formatter.fmt_year(2018, wildcard=True)
        self.assertEqual(actual, "*-2018")

    def test_fmt_month(self):
        # Test without wildcard
        actual = self.formatter.fmt_month(2018, 4, wildcard=False)
        self.assertEqual(actual, "04-2018")
        # Test with wildcard
        actual = self.formatter.fmt_month(2018, 4, wildcard=True)
        self.assertEqual(actual, "*-04-2018")

    def test_fmt_day(self):
        actual = self.formatter.fmt_day(2018, 4, 20)
        self.assertEqual(actual, "20-04-2018")

    def test_sep(self):
        formatter = self.module(sep=".")
        actual_year = formatter.fmt_year(2018, wildcard=False)
        actual_month = formatter.fmt_month(2018, 4, wildcard=False)
        actual_day = formatter.fmt_day(2018, 4, 20)
        self.assertEqual(actual_year, "2018")
        self.assertEqual(actual_month, "04.2018")
        self.assertEqual(actual_day, "20.04.2018")

        for sep in formatters.INVALID_SEP_CHARS:
            with self.assertRaises(ValueError):
                self.module(sep)


class TestMiddleEndianDateFormatter(unittest.TestCase):

    def setUp(self):
        self.module = formatters.MiddleEndianDateFormatter
        self.formatter = self.module()

    def test_fmt_year(self):
        # Test without wildcard
        actual = self.formatter.fmt_year(2018, wildcard=False)
        self.assertEqual(actual, "2018")
        # Test with wildcard
        actual = self.formatter.fmt_year(2018, wildcard=True)
        self.assertEqual(actual, "*-2018")

    def test_fmt_month(self):
        # Test without wildcard
        actual = self.formatter.fmt_month(2018, 4, wildcard=False)
        self.assertEqual(actual, "04-2018")
        # Test with wildcard
        actual = self.formatter.fmt_month(2018, 4, wildcard=True)
        self.assertEqual(actual, "04-*-2018")

    def test_fmt_day(self):
        actual = self.formatter.fmt_day(2018, 4, 20)
        self.assertEqual(actual, "04-20-2018")

    def test_sep(self):
        formatter = self.module(sep=".")
        actual_year = formatter.fmt_year(2018, wildcard=False)
        actual_month = formatter.fmt_month(2018, 4, wildcard=False)
        actual_day = formatter.fmt_day(2018, 4, 20)
        self.assertEqual(actual_year, "2018")
        self.assertEqual(actual_month, "04.2018")
        self.assertEqual(actual_day, "04.20.2018")

        for sep in formatters.INVALID_SEP_CHARS:
            with self.assertRaises(ValueError):
                self.module(sep)


if __name__ == "__main__":
    unittest.main()


tests/test_partitioning.py
import datetime
import unittest

from elasticsearch_partition import formatters, partitioning


class _PyDateFormatter(formatters.DateFormatter):

    def fmt_year(self, year, wildcard):
        date = datetime.date(year=year, month=1, day=1)
        if wildcard:
            return date.strftime("%Y{sep}*".format(sep=self.sep))
        return date.strftime("%Y")

    def fmt_month(self, year, month, wildcard):
        date = datetime.date(year=year, month=month, day=1)
        if wildcard:
            return date.strftime("%Y{sep}%m{sep}*".format(sep=self.sep))
        return date.strftime("%Y{sep}%m".format(sep=self.sep))

    def fmt_day(self, year, month, day):
        date = datetime.date(year, month, day)
        return date.strftime("%Y{sep}%m{sep}%d".format(sep=self.sep))


class TestRangePartition(unittest.TestCase):

    def setUp(self):
        self.cls = partitioning.RangePartition
        self.since = datetime.date(2014, 9, 27)
        self.until = datetime.date(2018, 2, 4)

    def test_partition_by_day(self):
        partition = self.cls(frequency=partitioning.DAY)
        expected = [
            "logs-2014-09-27",
            "logs-2014-09-28",
            "logs-2014-09-29",
            "logs-2014-09-30",
            "logs-2014-10-*",
            "logs-2014-11-*",
            "logs-2014-12-*",
            "logs-2015-*",
            "logs-2016-*",
            "logs-2017-*",
            "logs-2018-01-*",
            "logs-2018-02-01",
            "logs-2018-02-02",
            "logs-2018-02-03",
            "logs-2018-02-04",
        ]
        actual = partition("logs-*", self.since, self.until)
        self.assertListEqual(actual, expected)

    def test_partition_by_day_only_since(self):
        partition = self.cls(
            frequency=partitioning.DAY,
            now_func=lambda: datetime.date(2018, 7, 4)
        )
        expected = [
            "logs-2014-09-27",
            "logs-2014-09-28",
            "logs-2014-09-29",
            "logs-2014-09-30",
            "logs-2014-10-*",
            "logs-2014-11-*",
            "logs-2014-12-*",
            "logs-2015-*",
            "logs-2016-*",
            "logs-2017-*",
            "logs-2018-01-*",
            "logs-2018-02-*",
            "logs-2018-03-*",
            "logs-2018-04-*",
            "logs-2018-05-*",
            "logs-2018-06-*",
            "logs-2018-07-01",
            "logs-2018-07-02",
            "logs-2018-07-03",
            "logs-2018-07-04",
        ]
        actual = partition("logs-*", self.since)
        self.assertSequenceEqual(actual, expected)

    def test_partition_by_day_only_until(self):
        partition = self.cls(
            frequency=partitioning.DAY,
            now_func=lambda: datetime.date(2018, 7, 4)
        )
        expected = [
            "-logs-2018-02-04",
            "-logs-2018-02-05",
            "-logs-2018-02-06",
            "-logs-2018-02-07",
            "-logs-2018-02-08",
            "-logs-2018-02-09",
            "-logs-2018-02-10",
            "-logs-2018-02-11",
            "-logs-2018-02-12",
            "-logs-2018-02-13",
            "-logs-2018-02-14",
            "-logs-2018-02-15",
            "-logs-2018-02-16",
            "-logs-2018-02-17",
            "-logs-2018-02-18",
            "-logs-2018-02-19",
            "-logs-2018-02-20",
            "-logs-2018-02-21",
            "-logs-2018-02-22",
            "-logs-2018-02-23",
            "-logs-2018-02-24",
            "-logs-2018-02-25",
            "-logs-2018-02-26",
            "-logs-2018-02-27",
            "-logs-2018-02-28",
            "-logs-2018-03-*",
            "-logs-2018-04-*",
            "-logs-2018-05-*",
            "-logs-2018-06-*",
            "-logs-2018-07-01",
            "-logs-2018-07-02",
            "-logs-2018-07-03",
            "-logs-2018-07-04",
            "logs-*"
        ]
        actual = partition("logs-*", until=self.until)
        self.assertSequenceEqual(actual, expected)

    def test_partition_by_month(self):
        partition = self.cls(frequency=partitioning.MONTH)
        expected = [
            "logs-2014-09",
            "logs-2014-10",
            "logs-2014-11",
            "logs-2014-12",
            "logs-2015-*",
            "logs-2016-*",
            "logs-2017-*",
            "logs-2018-01",
            "logs-2018-02",
        ]
        actual = partition("logs-*", self.since, self.until)
        self.assertListEqual(actual, expected)

    def test_partition_by_month_only_since(self):
        partition = self.cls(
            frequency=partitioning.MONTH,
            now_func=lambda: datetime.date(2018, 7, 4)
        )
        expected = [
            "logs-2014-09",
            "logs-2014-10",
            "logs-2014-11",
            "logs-2014-12",
            "logs-2015-*",
            "logs-2016-*",
            "logs-2017-*",
            "logs-2018-01",
            "logs-2018-02",
            "logs-2018-03",
            "logs-2018-04",
            "logs-2018-05",
            "logs-2018-06",
            "logs-2018-07",
        ]
        actual = partition("logs-*", self.since)
        self.assertListEqual(actual, expected)

    def test_partition_by_month_only_until(self):
        partition = self.cls(
            frequency=partitioning.MONTH,
            now_func=lambda: datetime.date(2018, 7, 4)
        )
        expected = [
            "-logs-2018-02",
            "-logs-2018-03",
            "-logs-2018-04",
            "-logs-2018-05",
            "-logs-2018-06",
            "-logs-2018-07",
            "logs-*"
        ]
        actual = partition("logs-*", until=self.until)
        self.assertListEqual(actual, expected)

    def test_partition_by_year(self):
        partition = self.cls(frequency=partitioning.YEAR)
        expected = [
            "logs-2014",
            "logs-2015",
            "logs-2016",
            "logs-2017",
            "logs-2018",
        ]
        actual = partition("logs-*", self.since, self.until)
        self.assertListEqual(actual, expected)

    def test_partition_by_same_year(self):
        partition = self.cls(frequency=partitioning.YEAR)
        expected = ["logs-2018"]
        actual = partition(
            "logs-*",
            datetime.date(2018, 7, 4),
            datetime.date(2018, 8, 9)
        )
        self.assertListEqual(actual, expected)

    def test_partition_by_year_only_since(self):
        partition = self.cls(
            frequency=partitioning.YEAR,
            now_func=lambda: datetime.date(2018, 7, 4)
        )
        expected = [
            "logs-2014",
            "logs-2015",
            "logs-2016",
            "logs-2017",
            "logs-2018",
        ]
        actual = partition("logs-*", self.since)
        self.assertListEqual(actual, expected)

    def test_partition_by_year_only_until(self):
        partition = self.cls(
            frequency=partitioning.YEAR,
            now_func=lambda: datetime.date(2018, 7, 4)
        )
        expected = ["logs-*"]
        actual = partition("logs-*", until=self.until)
        self.assertListEqual(actual, expected)

    def test_partition_with_custom_formatter(self):
        partition = self.cls(
            frequency=partitioning.DAY,
            formatter=formatters.MiddleEndianDateFormatter(),
            escape="@"
        )
        expected = [
            "logs-09-27-2014",
            "logs-09-28-2014",
            "logs-09-29-2014",
            "logs-09-30-2014",
            "logs-10-*-2014",
            "logs-11-*-2014",
            "logs-12-*-2014",
            "logs-*-2015",
            "logs-*-2016",
            "logs-*-2017",
            "logs-01-*-2018",
            "logs-02-01-2018",
            "logs-02-02-2018",
            "logs-02-03-2018",
            "logs-02-04-2018",
        ]
        actual = partition("logs-@", self.since, self.until)
        self.assertListEqual(actual, expected)

    def test_partition_with_custom_py_formatter(self):
        partition = self.cls(
            frequency=partitioning.DAY,
            formatter=_PyDateFormatter(),
        )
        expected = [
            "logs-2014-09-27",
            "logs-2014-09-28",
            "logs-2014-09-29",
            "logs-2014-09-30",
            "logs-2014-10-*",
            "logs-2014-11-*",
            "logs-2014-12-*",
            "logs-2015-*",
            "logs-2016-*",
            "logs-2017-*",
            "logs-2018-01-*",
            "logs-2018-02-01",
            "logs-2018-02-02",
            "logs-2018-02-03",
            "logs-2018-02-04",
        ]
        actual = partition("logs-*", self.since, self.until)
        self.assertListEqual(actual, expected)

    def test_partition_with_wrong_pattern(self):
        partition = self.cls(escape="@")
        with self.assertRaises(ValueError) as ctx:
            partition("logs-*")

        self.assertEqual(
            "Index pattern 'logs-*' doesn't contain a special character '@'",
            str(ctx.exception)
        )

    def test_partition_without_tw_parameters(self):
        partition = self.cls()
        with self.assertRaises(ValueError) as ctx:
            partition("logs-*")

        self.assertEqual(
            "You should use 'since' or 'until' for searching by "
            "partitioning index",
            str(ctx.exception)
        )


if __name__ == "__main__":
    unittest.main()


scripts/benchmark.py
from datetime import date
import timeit


setup_template = """
import datetime
from elasticsearch_partition import partition

since = {since!r}
until = {until!r}
"""


def sample(name, number=100000, **kwargs):
    result = timeit.timeit(
        "partition('logs-*', since, until)",
        setup=setup_template.format(**kwargs),
        number=number
    )

    print("\nBenchmark: {}".format(name))

    for key in sorted(kwargs.keys()):
        print("{0}: {1}".format(key, kwargs[key]))

    usec = (result / number) * 1e6
    report = "{0} loops, {1:.2f} usec per loop".format(number, usec)
    print("{0}\n{1}\n".format("-" * (len(report)), report))


if __name__ == "__main__":
    sample("a", since=date(1970, 7, 10), until=date(2018, 8, 26))
    sample("b", since=date(2001, 2, 25), until=date(2018, 5, 3))
    sample("c", since=date(2012, 8, 15), until=date(2018, 3, 20))
    sample("d", since=date(2016, 10, 12), until=date(2018, 3, 20))
    sample("e", since=date(2018, 3, 1), until=date(2018, 3, 7))


elasticsearch_partition/partitioning.py
"""This module provide range partition method for your Elasticsearch
indexes. If you have a logic index partition you could use this module
for creating and searching through indexes by time range.

A partition is a division of a logical database or its constituent
elements into distinct independent parts. Database partitioning is
normally done for manageability, performance or availability reasons,
or for load balancing.
"""

from calendar import monthrange
import datetime
from itertools import groupby

from .formatters import BigEndianDateFormatter, DateFormatter


__all__ = ("RangePartitioning", "TimeWindow", "partition")


# Public constants for convenience
YEAR = 1
MONTH = 2
DAY = 3

_MAXMONTH = 12


def _group_by(iterable, key):
    """
    Group By implementation.

    :param Iterator iterable: Iterable sequesnce of date objects
    :param callable key: Function for getting attributes
    :yield (str, Iterator): Key with grouped values
    """
    for key, values in groupby(iterable, key=key):
        yield key, values


def _group_by_year(iterable):
    """
    Group and returns iterable sequence by year.

    :param Iterator iterable: Iterable sequesnce of date objects
    :return (year: str, Iterator): Grouped dates by year
    """
    return _group_by(iterable, key=lambda x: x.year)


def _group_by_month(iterable):
    """
    Group and returns iterable sequence by month.

    :param Iterator iterable: Iterable sequesnce of date objects
    :return (month: str, Iterator): Grouped dates by month
    """
    return _group_by(iterable, key=lambda x: x.month)


def _exclude(pattern):
    """
    Change index name to Elasticsearch exclude style.

    :param str pattern: Index name with a specified symbol
    :rtype: str
    """
    return "-" + pattern


class TimeWindow:
    __slots__ = ("since", "until")

    def __init__(self, since, until):
        if not isinstance(since, datetime.date):
            raise TypeError("The 'since' parameter must be a 'date' type")

        if not isinstance(until, datetime.date):
            raise TypeError("The 'until' parameter must be a 'date' type")

        if since > until:
            raise ValueError("'since' can't be more than 'until'")

        self.since = since
        self.until = until

    @property
    def delta_years(self):
        """
        Returns delta in years.
        """
        return self.until.year - self.since.year

    @property
    def delta_days(self):
        """
        Returns delta in days.
        """
        return (self.until - self.since).days

    def __iter__(self):
        """
        Iterate over time periods by day.
        """
        curren_date = self.since
        while curren_date <= self.until:
            yield curren_date
            curren_date += datetime.timedelta(days=1)


class RangePartitioning:
    """
    Range partitioning maps data to partitions based on ranges of values
    of the partitioning index name that you establish for each partition.
    It is the most common type of partitioning and is often used with dates.

    For an index with a date prefix as the partitioning key, the January-2018
    partition would has name with partitioning prefix from '2018-01-01'
    to '2018-01-31'.
    """

    def __init__(self,
                 frequency=DAY,
                 formatter=None,
                 escape="*",
                 now_func=None):
        """
        Initialize range partitioning instance.

        :param int frequency: Index partitioning frequency
        :param DateFormatter formatter: Formatter instance
        :param str escape: Special character which will be replaced on a date
        :param callable now_func: Get now date function
        """

        if formatter is None:
            self._formatter = BigEndianDateFormatter()
        else:
            if not isinstance(formatter, DateFormatter):
                raise TypeError(
                    "The 'formatter' must be an instance of 'DateFormatter'"
                )
            self._formatter = formatter

        self._frequency = frequency
        self._escape = escape
        self._now_func = now_func or self.__get_now_date

    def __get_now_date(self):
        return datetime.datetime.now().date()

    def partition(self, pattern, tm):
        """
        This method create and returns Elasticsearch partitioned indexes
        by date frequency.

        :param str pattern: Index name with a special character
        :param TimeWindow tm: Time window instance
        :rtype: list
        """

        def fmt_index(fmt_date):
            # Helper function for better index formatting
            return pattern.replace(self._escape, fmt_date)

        indexes = []
        for year, iterable in _group_by_year(tm):
            # Frequency by year
            if self._frequency == YEAR:
                fmt_date = self._formatter.fmt_year(year)
                indexes.append(pattern.replace(self._escape, fmt_date))
                continue

            # Temporary list for months and days indexes
            _indexes = []
            month_count = 0
            for month, iterable in _group_by_month(iterable):
                # Frequency by day
                if self._frequency > MONTH:
                    # Load all month days to a memory
                    days = list(iterable)
                    if len(days) != monthrange(year, month)[1]:
                        for date in days:
                            fmt_date = self._formatter.fmt_day(
                                date.year, date.month, date.day
                            )
                            _indexes.append(fmt_index(fmt_date))
                    else:
                        # Increase month counter and append formatted date
                        # to the temporary indexes list
                        month_count += 1
                        fmt_date = self._formatter.fmt_month(
                            year, month, wildcard=True
                        )
                        _indexes.append(fmt_index(fmt_date))

                # Frequency by month
                else:
                    month_count += 1
                    fmt_date = self._formatter.fmt_month(year, month)
                    _indexes.append(fmt_index(fmt_date))

            if month_count == _MAXMONTH:
                # Append formatted year with wildcard character if month
                # counter is full
                fmt_date = self._formatter.fmt_year(year,  wildcard=True)
                indexes.append(fmt_index(fmt_date))
            else:
                # Append temporary indexes to the main index list
                indexes.extend(_indexes)

        return indexes

    def __call__(self, pattern, since=None, until=None):
        """
        Returns Elasticsearch partitioned indexes by date range.

        :param str pattern: Index name with a special character
        :param datetime since: Since partitioning date
        :param datetime until: Until partitioning date
        """
        if self._escape not in pattern:
            raise ValueError("Index pattern '%s' doesn't contain a special "
                             "character '%s'" % (pattern, self._escape))

        if not since and not until:
            raise ValueError("You should use 'since' or 'until' for "
                             "searching by partitioning index")

        if not since:
            tm = TimeWindow(until, self._now_func())
            if self._frequency == YEAR and tm.delta_years < 1:
                indexes = [pattern]
            else:
                indexes = self.partition(_exclude(pattern), tm)
                indexes.append(pattern)
        else:
            tm = TimeWindow(since, until or self._now_func())
            indexes = self.partition(pattern, tm)

        return indexes


# Default Elasticsearch partitioning by day with big-endian date formatter
partition = RangePartitioning()


elasticsearch_partition/formatters.py
from abc import ABCMeta, abstractmethod
import datetime


INVALID_SEP_CHARS = ["\\", "/", "*", "?", "\"", "<", ">", "|", " ", ","]

# Abstract metaclass with py2/py3 support
Abstract = ABCMeta("Abstract", (object,), {})


class DateFormatter(Abstract):
    """
    Abstract date formatter class.
    """

    def __init__(self, sep="-"):
        """
        Accepts separation character and initialize specified
        date formatter.
        """
        if sep in INVALID_SEP_CHARS:
            raise ValueError(
                "The separation character '%s' is not valid" % sep
            )
        self._sep = sep

    @abstractmethod
    def fmt_year(self, year, wildcard=False):
        pass

    @abstractmethod
    def fmt_month(self, year, month, wildcard=False):
        pass

    @abstractmethod
    def fmt_day(self, year, month, day):
        pass


class BigEndianDateFormatter(DateFormatter):
    """
    In this format the most significant data item is written before
    lesser data items i.e. year before month before day.

    (year, month, day), e.g. 2018-04-22 or 2018.04.22 or 2018/04/22
    """

    def fmt_year(self, year, wildcard=False):
        """
        Format and returns year in big-endian order style.

        :param int year: Year
        :param bool wildcard: Use wildcard instead day and month
        :rtype: str
        """
        date = datetime.date(year=year, month=1, day=1)
        if wildcard:
            return date.strftime("%Y{sep}*".format(sep=self._sep))
        return date.strftime("%Y")

    def fmt_month(self, year, month, wildcard=False):
        """
        Format and returns month in big-endian order style.

        :param int year: Year
        :param int month: Month
        :param bool wildcard: Use wildcard instead day
        :rtype: str
        """
        date = datetime.date(year=year, month=month, day=1)
        if wildcard:
            return date.strftime("%Y{sep}%m{sep}*".format(sep=self._sep))
        return date.strftime("%Y{sep}%m".format(sep=self._sep))

    def fmt_day(self, year, month, day):
        """
        Format and returns day in big-endian order style.

        :param int year: Year
        :param int month: Month
        :param int day: Day
        :rtype: str
        """
        date = datetime.date(year, month, day)
        return date.strftime("%Y{sep}%m{sep}%d".format(sep=self._sep))


class LittleEndianDateFormatter(DateFormatter):
    """
    In this format the most significant data item is written after
    lesser data items i.e. day before month before year.

    (day, month, year), e.g. 22-04-2018 or 22.04.2018 or 22/04/2018
    """

    def fmt_year(self, year, wildcard=False):
        """
        Format and returns year in little-endian order style.

        :param int year: Year
        :param bool wildcard: Use wildcard instead day and month
        :rtype: str
        """
        date = datetime.date(year=year, month=1, day=1)
        if wildcard:
            return date.strftime("*{sep}%Y".format(sep=self._sep))
        return date.strftime("%Y".format(sep=self._sep))

    def fmt_month(self, year, month, wildcard=False):
        """
        Format and returns month in little-endian order style.

        :param int year: Year
        :param int month: Month
        :param bool wildcard: Use wildcard instead day
        :rtype: str
        """
        date = datetime.date(year=year, month=month, day=1)
        if wildcard:
            return date.strftime("*{sep}%m{sep}%Y".format(sep=self._sep))
        return date.strftime("%m{sep}%Y".format(sep=self._sep))

    def fmt_day(self, year, month, day):
        """
        Format and returns day in little-endian order style.

        :param int year: Year
        :param int month: Month
        :param int day: Day
        :rtype: str
        """
        date = datetime.date(year, month, day)
        return date.strftime("%d{sep}%m{sep}%Y".format(sep=self._sep))


class MiddleEndianDateFormatter(DateFormatter):
    """
    In this format the month data item is written before day before year.

    (month, day, year), e.g. 04-22-2018 or 04.22.2018 or 04/22/2018
    """

    def fmt_year(self, year, wildcard=False):
        """
        Format and returns year in middle-endian order style.

        :param int year: Year
        :param bool wildcard: Use wildcard instead day and month
        :rtype: str
        """
        date = datetime.date(year=year, month=1, day=1)
        if wildcard:
            return date.strftime("*{sep}%Y".format(sep=self._sep))
        return date.strftime("%Y")

    def fmt_month(self, year, month, wildcard=False):
        """
        Format and returns month in middle-endian order style.

        :param int year: Year
        :param int month: Month
        :param bool wildcard: Use wildcard instead day
        :rtype: str
        """
        date = datetime.date(year=year, month=month, day=1)
        if wildcard:
            return date.strftime("%m{sep}*{sep}%Y".format(sep=self._sep))
        return date.strftime("%m{sep}%Y".format(sep=self._sep))

    def fmt_day(self, year, month, day):
        """
        Format and returns day in middle-endian order style.

        :param int year: Year
        :param int month: Month
        :param int day: Day
        :rtype: str
        """
        date = datetime.date(year, month, day)
        return date.strftime("%m{sep}%d{sep}%Y".format(sep=self._sep))
